unit/goto-programs/restrict_function_pointers.cpp

```diff
@@ -88,6 +95,67 @@ void merge_restrictions_test()
   REQUIRE(fp2_restrictions.count("func1") == 1);
 }
 
+void get_function_pointer_by_name_restrictions_test()
+{
+  const std::string code = R"(
+    typedef void (*fp_t)(void);
+
+    fp_t fp;
+
+    void f();
+
+    void func(fp_t fp)
+    {
+      fp();
+    }
+
+    void main()
+    {
+      f(); // ignored
+
+      {
+        fp_t fp;
+        fp();
+      }
+
+      {
+        fp_t fp;
+        fp();
+      }
+
+      fp();
+      fp(); // second call to same function pointer
+    }
+  )";
+
+  goto_modelt goto_model = get_goto_model_from_c(code);
+
+  label_function_pointer_call_sites(goto_model);
+
+  const auto restrictions =
```

Comment:
:bulb: Ideally a unit test would test less components at once (e.g. this tests both the parsing of the options, and the restriction stuff) - if you split up the parsing, and have the result of the parsing being passed into the restrictions, the test would be simpler

Reply:
Now I understand, I think the test would have been easier to follow if was a couple of simpler examples:


```
void func1(fp_t fp)
    {
      fp();
    }
void func1(fp_t fp)
    {
      fp();
    }

TEST_CASE("Check function scope is correctly applied to variable names")
fp_restrictionst::get_function_pointer_by_name_restrictions(
      {"func1::fp/fa", "func1::fp/fa"},
      goto_model);
 const auto set1 = restrictions.at("func.function_pointer_call.1");
REQUIRE(restrictions == {{"func2.function_pointer_call.1", "fa"}, {"func2.function_pointer_call.1", "fb"})
...
``

then have a separate test case for checking the scope thing

Reply:
Split it up into several simpler cases.

Reply:
Much clearer thanks!